### DISCUSSION TITLE
Shell Extensions and Explorer Process?

### DISCUSSION THREAD

**Discussion by elthombre** [Squirrel/Squirrel.Windows#1803] [Q&A] [answered]
Hi,

We have a shell extension and a silent installer such as this might be a good choice. We know that we can leverage the regasm to register the Shell extension which should alleviate the need to restart the Explorer process.

Meanwhile, it is our understanding that we must sill restart the Explorer process on uninstall and, presently, looking at the Process class to handle this.

We are new to the Squirrel installer and were wondering if Squirrel can handle these specific requirements?

Thanks

**Comment by robmen** [comment, marked as answer]
Squirrel is not designed for this use case.

**Reply by elthombre** [reply]
Thanks

**Comment by anaisbetts** [comment]
You can technically do it but I would tend to agree with @robmen (though tbh, killing Explorer for any reason already puts you into "Gross hack" territory so I guess you may as well lol)

**Reply by elthombre** [reply]
not exactly - Shell extensions require a restart of the Explorer process.

**Reply by caesay** [reply]
The proper/usual way to uninstall a shell extension (or any other files that get locked by running processes) is to unregister them and mark them for deletion at the next reboot. Not to kill vital windows processes that are being used. In either case, this is well outside what is supported by Squirrel.

**Reply by elthombre** [reply]
not 'kill' but restart / relaunch. and when we say 'proper', to some, it is a less than ideal user experience to force a reboot especially if there is an alternative

**Reply by anaisbetts** [reply]
@caesay Unfortunately that's not practical for many shell extensions (if you delete all of the surrounding files of a shell extension but leave the actual DLL, nothing good is going to happen!)  Many Shell extensions have historically unfortunately had to resort to this workaround (regardless of install technology). 

The most "proper" way to do this is to uninstall the shell extension registry only, demand a reboot, then on next reboot do the full uninstall. But however disruptive killing Explorer is, forcing a reboot is absolutely more disruptive!

**Reply by caesay** [reply]
No need to debate further but I don't follow the logic here. Unregister your extensions via registry, your com servers etc. Don't delete anything. Queue everything to be cleaned up at next boot, and prompt the user to restart now or later, at the user's convenience. No need to force anything at all. Killing explorer or dwm will often necessitate a reboot anyways; because many applications do not recover well from this.

**Reply by elthombre** [reply]
the problem is that it is *not* convenient. For the user, it is not convenient to reboot. Therein lay the trouble.

**Reply by caesay** [reply]
Exactly. Which is why you do not kill / restart critical processes (which can break things). And why you don't force a reboot. Let the user choose to restart "Later", and it will get cleaned up the next time they naturally restart their computer. That might be a few hours later, or days later. At their convenience.

**Reply by anaisbetts** [reply]
I think the real moral of the story is, Shell Extensions suck, don't write them 😅
